pkg/policy/k8s/cell.go

```diff
@@ -66,63 +65,63 @@ type PolicyWatcherParams struct {
+	params PolicyWatcherParams
+}
+
+func newPolicyResourcesWatcher(p PolicyWatcherParams) *PolicyResourcesWatcher {
 	if !p.ClientSet.IsEnabled() {
-		return // skip watcher if K8s is not enabled
+		return nil // skip watcher if K8s is not enabled
+	}
+
+	return &PolicyResourcesWatcher{
+		params: p,
 	}
+}
 
-	// We want to subscribe before the start hook is invoked in order to not miss
-	// any events
-	ctx, cancel := context.WithCancel(context.Background())
-	svcCacheNotifications := stream.ToChannel(ctx, p.ServiceCache.Notifications(),
-		stream.WithBufferSize(int(p.Config.K8sServiceCacheSize)))
-
-	p.Lifecycle.Append(cell.Hook{
-		OnStart: func(startCtx cell.HookContext) error {
-			policyManager, err := p.PolicyManager.Await(startCtx)
-			if err != nil {
-				return err
-			}
-
-			w := &PolicyWatcher{
-				log:                              p.Logger,
-				config:                           p.Config,
-				k8sResourceSynced:                p.K8sResourceSynced,
-				k8sAPIGroups:                     p.K8sAPIGroups,
-				svcCache:                         p.ServiceCache,
-				svcCacheNotifications:            svcCacheNotifications,
-				policyManager:                    policyManager,
-				CiliumNetworkPolicies:            p.CiliumNetworkPolicies,
-				CiliumClusterwideNetworkPolicies: p.CiliumClusterwideNetworkPolicies,
-				CiliumCIDRGroups:                 p.CiliumCIDRGroups,
-				NetworkPolicies:                  p.NetworkPolicies,
-
-				cnpCache:          make(map[resource.Key]*types.SlimCNP),
-				cidrGroupCache:    make(map[string]*cilium_v2_alpha1.CiliumCIDRGroup),
-				cidrGroupPolicies: make(map[resource.Key]struct{}),
-
-				toServicesPolicies: make(map[resource.Key]struct{}),
-				cnpByServiceID:     make(map[k8s.ServiceID]map[resource.Key]struct{}),
-			}
-
-			w.watchResources(ctx)
-
-			return nil
-		},
-		OnStop: func(cell.HookContext) error {
-			if cancel != nil {
-				cancel()
-			}
-			return nil
-		},
-	})
+// WatchK8sPolicyResources starts watching Kubernetes policy resources.
+// Needs to be called before K8sWatcher.InitK8sSubsystem.
+func (p *PolicyResourcesWatcher) WatchK8sPolicyResources(ctx context.Context, policyManager PolicyManager) {
+	// This function unfortunately cannot be started via the Hive lifecycle as
+	// there exists a circular dependency between this watcher and the Daemon:
+	// The constructor newDaemon cannot complete before all pre-existing
+	// K8s policy resources have been added via the PolicyManager
+	// (i.e. watchResources has observed the Sync event for CNP/CCNP/KNP).
+	// Because the PolicyManager interface itself is implemented by the Daemon
+	// struct, we have a circular dependency.
+
+	// In order to not miss any service events, we register a subscriber here,
+	// before the call to K8sWatcher.InitK8sSubsystem.
+	svcCacheNotifications := stream.ToChannel(ctx, p.params.ServiceCache.Notifications(),
+		stream.WithBufferSize(int(p.params.Config.K8sServiceCacheSize)))
+
+	w := &policyWatcher{
```

Comment:
nit: Perhaps we could use a ctor here and hide the semantic + opts at the call site?

Reply:
Thanks. I've moved the code to a separate constructor. Is this what you had in mind?